<VirtualFocus /> component

JulianNymark opened the pull request:
internal tooling (for now)

JulianNymark committed:
:construction: WIP on some dropdown prop API

JulianNymark committed:
:construction: WIP on some dropdown prop API

JulianNymark committed:
:recycle: rename dropdown -> autocomplete

JulianNymark committed:
:recycle: rename dropdown -> autocomplete

JulianNymark committed:
Merge branch 'main' into search-dropdown

JulianNymark committed:
:construction: WIP towards Autocomplete (inspiration from Menu)

JulianNymark committed:
:construction: WIP towards Autocomplete (inspiration from Menu)

JulianNymark committed:
:construction: WIP Autocomplete

JulianNymark committed:
:construction: WIP Autocomplete

JulianNymark committed:
:construction: wip Autocomplete

JulianNymark committed:
:construction: wip Autocomplete

JulianNymark committed:
:construction: wip Autocomplete

JulianNymark committed:
:construction: wip Autocomplete

JulianNymark committed:
:construction: working idea for context (get & set autocomplete)

const thinger = () => {

JulianNymark committed:
:construction: working idea for context (get & set autocomplete)

const thinger = () => {

JulianNymark committed:
:construction: pick property on Item

JulianNymark committed:
:construction: pick property on Item

JulianNymark committed:
:construction: enrich story/example of Autocomplete

show how easy it is to add icons / do something more complicated inside
the autocomplete suggestions list alongside a <Search>

JulianNymark committed:
:construction: enrich story/example of Autocomplete

show how easy it is to add icons / do something more complicated inside
the autocomplete suggestions list alongside a <Search>

JulianNymark committed:
:construction: wip

- fairly nicely working list of virtually focusable items
- API interface kept simple (moved pick to Anchor though, single "exit
  point")
- item has focus function now, which triggers when it gets virtual
  focus, the API should not distinguish between virtual and real focus
  was the thinking here, it's more an implementation detail.

JulianNymark committed:
:construction: wip

- fairly nicely working list of virtually focusable items
- API interface kept simple (moved pick to Anchor though, single "exit
  point")
- item has focus function now, which triggers when it gets virtual
  focus, the API should not distinguish between virtual and real focus
  was the thinking here, it's more an implementation detail.

JulianNymark committed:
:construction: WIP new story for Autocomplete

- showcase dropdown search autocomplete

JulianNymark committed:
:construction: WIP new story for Autocomplete

- showcase dropdown search autocomplete

KenAJoh commented on @navikt/core/react/src/form/autocomplete/utils.ts:
Think this could be implemented by the consumer since the focus-marking would differ based on implementation. If we add a `data`-attrb the user could easily do a simple css-snippet 
```css
[data-aksel-virtualfocus="true"] {...}
// or
[data-aksel-virtualfocus="true"] > .heading {...}
```

Could then also remove [@navikt/core/css/autocomplete.css](https://github.com/navikt/aksel/pull/2894/files#diff-58ecbea66fbd457e48c0b7bb1838c949b51f63aeb697fca1b0df8bed16957caf)

JulianNymark replied:
agreed! (it's not always a 3px blue border around the exact element getting focus)

KenAJoh commented on @navikt/core/react/src/form/search/Search.tsx:
Are the changes made to search and search.css meant to be added to this PR?

JulianNymark replied:
ah, I'll remove them now, not really relevant anymore, later maybe it can be added as using the Autocomplete component

KenAJoh commented on @navikt/core/react/src/util/hooks/descendants/descendant.ts:
```diff
@@ -9,6 +9,11 @@ export type DescendantOptions<T = object> = T & {
+    node_to_focus: HTMLElement,
+    node_to_blur?: HTMLElement,
+  ) => void;
+  handlePick?: () => void;
```
These should not be typed here, but rather when creating the descendant context. (Added a comment with howto )

KenAJoh commented on @navikt/core/react/src/form/autocomplete/Autocomplete.tsx:
```diff
@@ -0,0 +1,214 @@
+  useAutocompleteDescendantsContext,
+  useAutocompleteDescendants,
+  useAutocompleteDescendant,
+] = createDescendantContext<SlottedDivElementRef>();
```
Ref where to type the additional props
```suggestion
] = createDescendantContext<SlottedDivElementRef, {   handleVirtualOnFocus?: (
    node_to_focus: HTMLElement,
    node_to_blur?: HTMLElement,
  ) => void;
  handlePick?: () => void;
  }>();
```

KenAJoh commented on @navikt/core/react/src/form/autocomplete/Autocomplete.tsx:
```diff
@@ -0,0 +1,214 @@
+  const [virtualFocusIdx, setVirtualFocusIdx] = useState(0);
+  return (
+    <AutocompleteInternalContextProvider
+      {...{ virtualFocusIdx, setVirtualFocusIdx }}
```
```suggestion
      virtualFocusIdx={virtualFocusIdx}
      setVirtualFocusIdx={setVirtualFocusIdx}
```

KenAJoh commented on @navikt/core/react/src/form/autocomplete/Autocomplete.tsx:
```diff
@@ -0,0 +1,214 @@
+    <AutocompleteInternalContextProvider
+      {...{ virtualFocusIdx, setVirtualFocusIdx }}
+    >
+      <AutocompleteContextProvider {...{ value, setValue }}>
```
```suggestion
      <AutocompleteContextProvider value={value} setValue={setValue}>
```

KenAJoh commented on @navikt/core/react/src/form/autocomplete/Autocomplete.tsx:
```diff
@@ -0,0 +1,214 @@
+            }}
+          >
+            {children}
+          </div>
```
Not 100% on this one, but should we assume that VirtualFocus will always be controlles? We dont need to have an internal state for `value` and `setState`, but rather assume that the user handles the input values.

KenAJoh commented on @navikt/core/react/src/form/autocomplete/Autocomplete.tsx:
```diff
@@ -0,0 +1,214 @@
+export const Autocomplete = ({ children }: { children: React.ReactNode }) => {
+  const descendants = useAutocompleteDescendants();
+  const [value, setValue] = useState("");
+  const [virtualFocusIdx, setVirtualFocusIdx] = useState(0);
```
Not sure how to best handle this, but each time the value changes the index needs should be reset to first element (anchor?) like in combobox https://aksel.nav.no/komponenter/core/combobox

KenAJoh commented on @navikt/core/react/src/form/autocomplete/autocomplete.stories.tsx:
https://github.com/navikt/aksel/assets/26967723/c18bd8c7-23c8-49a8-8e8e-f45acc2d47f4

Not sure of the cause (or just mac-stuff), but when focusing with keyboard, then moving mouse again if often lags behind before activating. Might be caused by `onMouseOver` not re-activating since its already over 🤔 Maybe `onMouseMove` or `onMouseMoveCapture` solves it.

JulianNymark replied:
I was thinking mouseMove too, I guess it doesn't matter if it fires a lot of events on the same element repeatedly, it just felt wrong / inefficient 🤔

KenAJoh replied:
Will prob need to check performance, but as long as the functions have the ability to exit early without some expensive calulcations its prob fine.

KenAJoh replied:
Could example have a ref with "isMouseOver" that is set to true/false based on onMouseMove and onMouseLeave to exit early

JulianNymark replied:
Hmm... thinking about this a bit... is it a feature or a bug 🐛 ? 🤔  (google search seems to have it "instant" so they probably use mouseMove)...  I'll switch to it since it doesn't seem to cause any issues.

KenAJoh commented on @navikt/core/react/src/util/hooks/descendants/descendant.stories.tsx:
```diff
@@ -126,9 +126,11 @@ function Input() {
       onKeyDown={(event) => {
         if (event.key === "ArrowRight") {
+          event.preventDefault();
           descendants.next(index, false)?.node.focus();
```
With next (and prev) multiple instances i would argue that "loop" should be false to avoid it jumping from bottom -> start or start -> bottom with arrowDown/Up
```
descendants.next(..., false)
```

JulianNymark replied:
It seemed google search does the same behaviour (loops around), but maybe it's not that common actually 🤔 (Duckduckgo has the same, but only on their "main page search")

KenAJoh replied:
Could just be a prop, example combobox wants it off, but maybe a search wants it on 🤔

JulianNymark committed:
Update @navikt/core/react/src/form/autocomplete/Autocomplete.tsx

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

JulianNymark committed:
Update @navikt/core/react/src/form/autocomplete/Autocomplete.tsx

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

JulianNymark committed:
Update @navikt/core/react/src/form/autocomplete/Autocomplete.tsx

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

JulianNymark committed:
Update @navikt/core/react/src/form/autocomplete/Autocomplete.tsx

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

JulianNymark committed:
:fire: refactor class -> attr, remove css, use user provided <style> in story

JulianNymark committed:
:fire: refactor class -> attr, remove css, use user provided <style> in story

JulianNymark committed:
Merge branch 'search-dropdown' of github.com:navikt/aksel into search-dropdown

JulianNymark committed:
:fire: remove originial ideas for a prop on <Search>

JulianNymark committed:
:fire: remove originial ideas for a prop on <Search>

JulianNymark committed:
:recycle: refactor + use mouseMove over mouseOver

JulianNymark committed:
:recycle: refactor + use mouseMove over mouseOver

JulianNymark committed:
:bug: reset virtualfocus on input change (within Anchor)

JulianNymark committed:
:bug: reset virtualfocus on input change (within Anchor)

JulianNymark committed:
:sparkles: add prop 'loop' for Autocomplete

JulianNymark committed:
:sparkles: add prop 'loop' for Autocomplete

JulianNymark committed:
:truck: Autocomplete -> VirtualFocus

JulianNymark committed:
:truck: Autocomplete -> VirtualFocus

JulianNymark committed:
Merge branch 'search-dropdown' of github.com:navikt/aksel into search-dropdown

JulianNymark committed:
:truck: Autocomplete -> VirtualFocus

JulianNymark committed:
Merge branch 'main' into search-dropdown

JulianNymark committed:
:fire: remove unused CSS

JulianNymark committed:
:construction: add onActive, delete unused code

JulianNymark committed:
:construction: rename CSS classes

JulianNymark committed:
:recycle: simplify <VirtualFocus> impl

- use the descendants API > grabbing the DOM nodes...

JulianNymark committed:
Merge branch 'main' into search-dropdown

JulianNymark committed:
:fire: remove the aria-activedescendant responsibility from VirtualFocus

KenAJoh commented on aksel.nav.no/website/sanity/config.ts:
```diff
@@ -1,6 +1,6 @@
 export const SANITY_PROJECT_ID = "hnbe3yhs";
 export const SANITY_API_VERSION = "2021-10-21";
-export const SANITY_DATASET = "production";
+export const SANITY_DATASET = "development";
```
```suggestion
export const SANITY_DATASET = "production";
```

KenAJoh committed:
Update aksel.nav.no/website/sanity/config.ts

JulianNymark committed:
:fire: remove comments that no longer apply

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,213 @@
+      pick();
+    },
+    handleOnActive: () => {
+      setVirtualFocusIdx(0);
```
Move this to "handlePick" since we dont want to reset focus index when f.example hovering or using keyboard-navigation before selecting anything.

JulianNymark replied:
The thinking was that active == "visual", and pick == "do something". But the naming is probably poor.

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
General updates:
- Add forwardRef on components
- Add rest spread of props on components where needed
- use `composeEventhandler` in cases where `onKeyDown`, `onBlur`, `onFocus` ++ are used to allow user to also access them

JulianNymark replied:
This is mostly resolved, only issue left is whether to use `composeEventhandler`, I don't think it's needed in this instance, since currently the only event listener set by us is `onKeyDown` on the Item, and it's an empty placeholder to satisfy eslint more than anything else (maybe we could ignore the rule instead?). It'll get overridden by user supplied params (through `...rest`), and the Items are supposed to have most of their logic in the supplied `onActive` and `pick` functions instead, since they are to be virtually focused, and therefore not "really focusable via keyboard"? 🤔.

KenAJoh replied:
> maybe we could ignore the rule instead
Like this better than empty eventhandeler. 

As for `{...rest}` i can see the argument for not needing `composeEventhandler` in v1 👍 But in the instances where rest is added at the end of the component (`VirtualFocusItem`) one will end up breaking it by the user adding a `onClick` or some of the other attributes. My vote here would be to either move `...rest` to the top so that the component still works in this case.

JulianNymark committed:
:fire: remove biome.json changes from PR

JulianNymark committed:
:recycle: various PR changes

- homogenize the API (so both anchor and item have the sameish props,
  onActive, pick...)
- use useId() so multiple <VirtualFocus> on the same page get unique ids

JulianNymark committed:
:construction: + :recycle: testing changes in VO

- it's not great, weird behaviour on readouts of Items... (FF completely
  broken, sometimes reads "rgla" from "hourglass", other times nothing,
  Chrome is a little better)
- is this a decent API at all? is it overcomplicated?

JulianNymark committed:
Merge branch 'main' into search-dropdown

JulianNymark committed:
:label: type itemRole

JulianNymark committed:
:fire: remove empty file

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+   * The function that is run when the focused element
+   * is to be picked (eg. do an actual search, change route... etc)
+   */
+  pick: () => void;
```
```suggestion
  onSelect: () => void;
```
Think maybe `onSelect` or `onPick` would be more in line with our system

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+   */
+  pick: () => void;
+  children: React.ReactNode;
+  itemRole?: React.AriaRole;
```
Since `role` is part of `React.HTMLAttributes<HTMLDivElement>`, we can skip re-assigning it and just use `role`

JulianNymark replied:
for containerRole, we should limit it to the subset of roles that use `aria-activedescendant` 🤔 

``` tsx
containerRole:
    | "combobox"
    | "grid"
    | "listbox"
    | "menu"
    | "menubar"
    | "radiogroup"
    | "tree"
    | "treegrid"
    | "tablist";
```

KenAJoh replied:
We can do something like this: 
```
  role:
    | "combobox"
    | "grid"
    | "listbox"
    | "menu"
    | "menubar"
    | "radiogroup"
    | "tree"
    | "treegrid"
    | "tablist";
```
since the union is a subset of `role`, this will work while limiting the available options 🤞

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+        data-aksel-virtualfocus={virtualFocusIdx === index}
+        ref={mergedRefs}
+        tabIndex={-1}
+        onKeyDown={() => {}} // Visible, non-interactive elements with click handlers must have at least one keyboard listener
```
Can omit this warning with eslint expect error

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+    return (
+      <div
+        id={`virtualfocus-${uniqueId}-${index}`}
+        className="navds-virtualfocus-item"
```
```suggestion
```
Since the consumer will declare their own scoped `className` i guess this can be removed.

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+          if (curr) {
+            setVirtualFocusIdx(curr.index);
+            curr.handleOnActive();
+          }
```
Can use `index` from `const { register, descendants, index } = useVirtualFocusDescendant({` here to skip having to look it up in `descendants`
```suggestion
        onMouseMove={(event) => {
            setVirtualFocusIdx(index);
            onActive();
```

JulianNymark replied:
right, so self reference doesn't need to call the wrapped function inside its descendant item... but I for some reason had this idea of collapsing everything to the same pattern, despite it being "roundabout & inefficient". 🤔 ... but I don't think it saves much on confusion.

KenAJoh replied:
Yea these changes are purely for performance on my side since it could lead to some unwanted delays if the user swipes the mouse over the list and ends up looping over every item for each mouseOver.

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+          if (curr) {
+            curr.handlePick();
+          }
+        }}
```
We can skip doing any lookups here since we know what function that needs to be called ⚡ 
```suggestion
        onClick={(event) => {
            pick();
        }}
```

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+  return (
+    <div
+      id={`virtualfocus-${uniqueId}-content`}
+      className="navds-virtualfocus-content"
```
```suggestion
    {...rest}
```

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+  children,
+}: {
+  children: React.ReactNode;
+}) => {
```
Since it's a `div`, the type should extend `HTMLDivElement`, and do prop-spread ⚡ (And forwardRef can be added)

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+  return (
+    <div
+      id={`virtualfocus-${uniqueId}-${index}`}
+      className="navds-virtualfocus-anchor"
```
```suggestion
```

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+   * the Anchor that you would rather tab to directly instead
+   * of having to tab to the Anchor itself first.
+   *
+   * @default 0
```
```suggestion
   * @default -1
```

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,251 @@
+    | "radiogroup"
+    | "tree"
+    | "treegrid"
+    | "tablist";
```
Since `containerRole` is only used within Anchor, i think it can be moved to the anchor-props, renamed to `role` and just set as required.

JulianNymark committed:
:recycle: pr comments

- renaming to better match the rest of DS
- remove itemRole (just use role)
- eslint ignore

JulianNymark committed:
:zap: do not go via descendant API if self

JulianNymark committed:
:art: add forwardRef & ...rest to Content

JulianNymark committed:
:bug: fix lying prop doc comment, remove unused CSS class

JulianNymark committed:
:recycle: use <Slot /> (hoistmerge child)

JulianNymark committed:
:lipstick: fix broken CSS (class was removed)

JulianNymark committed:
:memo: update jsdoc

JulianNymark committed:
:truck: move role to Anchor (from VirtualFocus wrapper)

JulianNymark committed:
:art: remove filtering from broken example

JulianNymark committed:
:bug: Anchor should trigger virtualfocused.onSelect()

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,228 @@
+      virtualFocusIdx={virtualFocusIdx}
+      setVirtualFocusIdx={setVirtualFocusIdx}
+      loop={loop}
+      uniqueId={useId().replace(/:/g, "")}
```
Since we still support React v17, we can't use React built-in `useId`(React 18), have to use this instead
```
import { useId } from "../../util/hooks";
```

KenAJoh replied:
Can also remove `.replace(/:/g, "")` then i think

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,228 @@
+          }
+        }
+      }}
+      {...rest}
```
```suggestion
ref={mergedRefs}
      {...rest}
      id={`virtualfocus-${uniqueId}-${index}`}
      tabIndex={0}
      aria-owns={`virtualfocus-${uniqueId}-content`}
      aria-controls={`virtualfocus-${uniqueId}-content`}
      aria-activedescendant={`virtualfocus-${uniqueId}-${virtualFocusIdx}`}
      onKeyDown={composeEventHandlers(rest.onKeyDown, (event) => {
        if (event.key === "ArrowDown") {
          event.preventDefault();
          const to_focus_descendant = descendants.next(virtualFocusIdx, loop);
          if (to_focus_descendant) {
            to_focus_descendant.handleOnActive();
          }
        } else if (event.key === "ArrowUp") {
          event.preventDefault();
          const to_focus_descendant = descendants.prev(virtualFocusIdx, loop);
          if (to_focus_descendant) {
            to_focus_descendant.handleOnActive();
          }
        } else if (event.key === "Enter") {
          const curr = descendants.item(virtualFocusIdx);
          if (curr?.handleOnSelect) {
            curr.handleOnSelect();
          }
        }
      })}
```

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,228 @@
+>(({ children, ...rest }, ref) => {
+  const { uniqueId } = useVirtualFocusInternalContext();
+  return (
+    <div ref={ref} id={`virtualfocus-${uniqueId}-content`} {...rest}>
```
```suggestion
    <div ref={ref} {...rest} id={`virtualfocus-${uniqueId}-content`}>
```

KenAJoh commented on @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,228 @@
+          setVirtualFocusIdx(index);
+          onActive();
+        }}
+        {...rest}
```
```suggestion
ref={mergedRefs}
        {...rest}
        id={`virtualfocus-${uniqueId}-${index}`}
        data-aksel-virtualfocus={virtualFocusIdx === index}
        tabIndex={-1}
        onClick={composeEventHandlers(rest.onClick, () => {
          onSelect();
        })}
        onMouseMove={composeEventHandlers(rest.onMouseMove, () => {
          setVirtualFocusIdx(index);
          onActive();
        })}
```

KenAJoh reviewed (dismissed):
Since this isn't an actual form-component, there might be a more fitting location for the code than under `/form`

JulianNymark committed:
:bug: use react v17 shimmed useId()

JulianNymark committed:
Update @navikt/core/react/src/form/virtualfocus/VirtualFocus.tsx

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

JulianNymark committed:
:construction: import

JulianNymark committed:
:art: improve story

JulianNymark committed:
:bug: do not let consumer overwrite certain VirtualBox implementation

JulianNymark committed:
:art: update story

JulianNymark committed:
Merge branch 'main' into search-dropdown

JulianNymark committed:
Merge branch 'main' into search-dropdown

JulianNymark committed:
:truck: move virtualfocus to utils

JulianNymark committed:
Merge branch 'main' into search-dropdown

JulianNymark committed:
Merge branch 'main' into search-dropdown

JulianNymark committed:
Merge branch 'main' into search-dropdown

JulianNymark committed:
Merge branch 'main' into search-dropdown

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,223 @@
+  children: React.ReactNode;
+  /**
+   * Whether to cause focus to loop around when it hits the first or last element
+   * @default true
```
```suggestion
   * @default false
```

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,223 @@
+  loop?: boolean;
+};
+
+export const VirtualFocus = ({ children, loop = false }: VirtualFocusProps) => {
```
Consider adding a description of each component.

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/VirtualFocus.tsx:
```diff
@@ -0,0 +1,223 @@
+export interface VirtualFocusContentProps
+  extends React.HTMLAttributes<HTMLDivElement> {}
+
+export const VirtualFocusContent = forwardRef<
```
Consider adding a "separator" between each component, so that it's easier to "scan" where one component ends and the next one starts. (Like here: https://github.com/navikt/aksel/pull/2906/files)

JulianNymark replied:
I split this into separate files in `parts/`, hopefully this is what you meant? (that's what is being done in link)

JulianNymark committed:
:art: update to latest RovingFocus.tsx, tweaks to stories

JulianNymark committed:
Update @navikt/core/react/src/util/virtualfocus/VirtualFocus.tsx

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

JulianNymark committed:
:art: improve docs (jsdoc), add Simple story

JulianNymark committed:
:truck: split up monster component into /parts

KenAJoh commented on @navikt/core/react/src/util/virtualfocus/parts/VirtualFocusContent.tsx:
```diff
@@ -0,0 +1,17 @@
+import React, { forwardRef } from "react";
+import { useVirtualFocusInternalContext } from "../VirtualFocus";
+
```
<img width="622" alt="Screenshot 2024-06-27 at 13 39 02" src="https://github.com/navikt/aksel/assets/26967723/facb52c2-2474-42ec-b9bc-4e3b3a1168ab">

Github detects cyclic imports now 🤯

JulianNymark replied:
neat! 🚀

JulianNymark replied:
or actually, I think it's just forwarding the string output from our lint step? 🤔 maybe.

JulianNymark committed:
:bug: pull out dependency to remove cyclic dependency

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/parts/VirtualFocusAnchor.tsx:
```diff
@@ -0,0 +1,104 @@
+   * its first child, you most likely want to keep this as `0`.
+   * @default 0
+   */
+  tabIndex?: number;
```
Seems like any `tabIndex` from the "outside" would be overridden on line 76...?

JulianNymark replied:
that's correct, good catch!

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/parts/VirtualFocusAnchor.tsx:
```diff
@@ -0,0 +1,104 @@
+      tabIndex={0}
+      aria-owns={`virtualfocus-${uniqueId}-content`}
+      aria-controls={`virtualfocus-${uniqueId}-content`}
+      aria-activedescendant={`virtualfocus-${uniqueId}-${virtualFocusIdx}`}
```
I'm not an expert on this, but does it make sense to set `aria-activedescendant` to itself when it has focus? 🤔

JulianNymark replied:
> The value of aria-activedescendant refers to an owned element of the controlled element. For example, in a combobox, focus may remain on the combobox while the value of aria-activedescendant on the combobox element refers to a descendant of a popup listbox that is controlled by the combobox.

It might make sense... the above from MDN https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-activedescendant, says it refers to an owned element of the controlled element, and an element always owns itself? 🤔. 

I want the virtual focus to be able to "go back to the Anchor" since that's the input element itself, but maybe this is the wrong way to go about it? 🤔 It was a "simple" solution however, and it seemed to work, but I should test this more (and probably read the specs more too). If the solution breaks when we start to use this component later on (I think we have some good tests already for Combobox + we will test the impl more thoroughly than the "simple story tests" I have come up with for this component so far).

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/parts/VirtualFocusAnchor.tsx:
```diff
@@ -0,0 +1,104 @@
+} from "../Context";
+
+export interface VirtualFocusAnchorProps
+  extends React.HTMLAttributes<HTMLDivElement> {
```
Should `id` be omitted?

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/RovingFocus.tsx:
This file doesn't seem to be in use anywhere.

JulianNymark replied:
You're right 😳, the usage was removed very early on, and I just assumed it was associated with the looping focus behaviour... but it's not. Removing 🔥

JulianNymark committed:
:fire: RovingFocus.tsx

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/parts/VirtualFocusItem.tsx:
```diff
@@ -0,0 +1,62 @@
+      },
+      handleOnSelect: () => {
+        onSelect();
+      },
```
Maybe you can just do:
```suggestion
      handleOnSelect: onSelect,
```

JulianNymark replied:
Yeah, there's no parameters here to think about, cleaner code 👏

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/parts/VirtualFocusItem.tsx:
```diff
@@ -0,0 +1,62 @@
+} from "../Context";
+
+export interface VirtualFocusItemProps
+  extends React.HTMLAttributes<HTMLDivElement> {
```
Should we omit `id` and `tabIndex`?

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/parts/VirtualFocusContent.tsx:
```diff
@@ -0,0 +1,17 @@
+import { useVirtualFocusInternalContext } from "../Context";
+
+export interface VirtualFocusContentProps
+  extends React.HTMLAttributes<HTMLDivElement> {}
```
Should we omit the `id`?

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/virtualfocus.stories.tsx:
Since this is just a util, should we disable Chromatic?

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/virtualfocus.stories.tsx:
```diff
@@ -0,0 +1,381 @@
+          value={value}
+          onChange={(current_input) => {
+            setValue(current_input);
+            to_focus?.handleOnActive();
```
What is this for? `to_focus?.handleOnActive();`

JulianNymark replied:
I don't remember currently. Might be old unused code that wasn't removed, or copy pasted & forgotten to modify. Either way, it seems to be non-impactful / not useful.

Tested without it, and it changes no behaviour. removed 🔥

HalvorHaugan commented on @navikt/core/react/src/util/virtualfocus/virtualfocus.stories.tsx:
```diff
@@ -0,0 +1,381 @@
+          <VirtualFocus.Content>
+            <VStack>
+              <MyItem setValue={setValue} icon={<CloudIcon aria-hidden />}>
+                {"cloud "}
```
Why is there a whitespace after each item?

JulianNymark replied:
ah, this was intentional, only to make the demo for the autocomple / dropdown example to feel a bit more convenient, saving a keystroke in case you wanted to add another search term. But it doesn't really make sense regardless, since there's no "real usecase" behind this example search box that necessitates adding several terms.

JulianNymark committed:
:bug: do not ignore external prop

JulianNymark committed:
:art: simplify function calls + reorder for consistency

JulianNymark committed:
:fire: remove unused code

JulianNymark committed:
:art: disable chromatic for internal tool stories

JulianNymark committed:
:label: Omit<> props that will get ignored/overridden

JulianNymark committed:
:label: Omit<> props that will get ignored/overridden

JulianNymark committed:
:label: use satisfies

JulianNymark committed:
Merge branch 'main' into search-dropdown
